File: Array/1.Easy/02.Second_largest_element_in_array.cpp
/*
QUESTION:-
Given an array Arr of size N, print second largest distinct element from an array.

Example:

Input:
N = 6
Arr[] = {12, 35, 1, 10, 34, 1}
Output: 34
Explanation: The largest element of the
array is 35 and the second largest element
is 34.
*/

/*
APPROACH
-> If the current element is larger than 'large' then update second_large and large variables
-> Else if the current element is larger than 'second_large' then we update the variable second_large.
-> Once we traverse the entire array, we would find the second largest element in the variable second_large.
*/

#include <bits/stdc++.h>
using namespace std;

// CODE:-
int print2largest(int arr[], int n)
{
    int prev = -1, curr = arr[0];
    for (int i = 1; i < n; i++)
    {
        if (arr[i] > curr)
        {
            prev = curr;
            curr = arr[i];
        }
        else if (arr[i] > prev && arr[i] != curr)
            prev = arr[i];
    }
    return prev;
}

int main() {
    // Test case 1
    cout << "Test 1:" << endl;
    int N1 = 6;
    int arr1[] = {12, 35, 1, 10, 34, 1};
    int result1 = print2largest(arr1, N1);
    cout << "Output: " << result1 << endl;
    cout << "Expected: 34" << endl;
    cout << endl;

    return 0;
}

// TIME COMPLEXITY = O(N)

File: 01.Arrays/1.Easy/02.Second_largest_element_in_array.py
# QUESTION:-
# Given an array Arr of size N, print second largest distinct element from an array.
# Example: Input: N = 6, Arr[] = {12, 35, 1, 10, 34, 1}, Output: 34

# APPROACH:-
# If current element is larger than 'large' then update second_large and large
# Else if current element is larger than 'second_large' then update second_large
# Once we traverse entire array, second_large contains the answer

# CODE:-
def print2largest(arr, n):
    """Find the second largest distinct element"""
    prev = -1
    curr = arr[0]
    for i in range(1, n):
        if arr[i] > curr:
            prev = curr
            curr = arr[i]
        elif arr[i] > prev and arr[i] != curr:
            prev = arr[i]
    return prev


if __name__ == '__main__':
    # Test case 1
    print("Test 1:")
    N1 = 6
    arr1 = [12, 35, 1, 10, 34, 1]
    result1 = print2largest(arr1, N1)
    print(f"Output: {result1}")
    print(f"Expected: 34")

# TIME COMPLEXITY = O(N) 
